Refactor and document sorting dispatch

LilithHafner opened the pull request:
Before, sorting dispatch used an ad-hoc system. I won't describe it here.

In this PR, sorting dispatch is specified by a nested chain of algorithms. For example, `MissingOptimization(BoolOptimization(Small{10}(IEEEFloatOptimization(QuickSort))))` will attempt optimizations to handle `eltypes` that are unions with `Missing` and those that are `Bool`. Then it will dispatch to a small algorithm for vectors shorter than 10 elements, perform an optimization for eltypes that are `IEEFloat`s, and finally, perform `QuickSort`.

This is a major refactoring, and part of a larger plan specified [here](https://github.com/JuliaLang/julia/discussions/44876#discussioncomment-2768446). It should make future development more pleasant, improve the interactions with third party sorting libraries (e.g. SortingAlgorithms.jl), and facilitate future removal of many of the optimizations from base to put them in a stdlib.

For the most part, I attempted to limit observable changes to dispatch but I did separate missing optimizations from IEEE floating point optimizations. I kept a fair bit of code re-use, but they are fundamentally separate optimizations, and separating them closes #27781.

Further, I made the missing optimization call downstream algorithms with an abstract vector with eltype that does not include `Missing`. This allows that optimization to be combined with other optimizations like that for `Bool`s and the `UIntMappable` suite.

One big difference between `sort!` and the `_sort!` function that this PR defines is that `_sort!` returns a scratch space (if any) rather than the sorted vector. Right now, that features is only lightly used, but it makes it possible to re-use scratch space allocated within a sorting method without pre-allocating beforehand. This is nice because it is often not obvious whether a sorting method will need scratch space at all. This fixes #47538, though it is likely not the only possible way to fix that issue.

I ran several benchmarks, many of which report runtime improvements.

Benchmark results EDITED to keep up to date with the PR
```julia
# mixed eltype with by order
julia> @btime sort(x; by=x -> x isa Symbol ? (0, x) : (1, x)) setup=(x=[rand() < .5 ? randstring() : Symbol(randstring()) for _ in 1:100]);
  31.586 μs (6 allocations: 1.88 KiB)  =>  30.311 μs (6 allocations: 1.88 KiB)

# reverse sorting Vector{Missing} (was 2ns in a previous version of this PR)
julia> @btime sort!(x, rev=true) setup=(x=fill(missing, 100));
  161.491 ns (3 allocations: 144 bytes)  =>  85.620 ns (0 allocations: 0 bytes)

# very small is still fast
julia> @btime sort(x) setup=(x=rand(2));
  56.484 ns (1 allocation: 80 bytes)  =>  52.098 ns (1 allocation: 80 bytes)

# sorting Vector{Missing} is O(1)
julia> @btime sort!(x) setup=(x=fill(missing, 100));
  163.465 ns (3 allocations: 144 bytes)  =>  4.329 ns (0 allocations: 0 bytes)

# sort(rand(1000)) is comparable
julia> @btime sort(x) setup=(x=rand(1000));
  16.383 μs (4 allocations: 18.09 KiB)  =>  16.369 μs (3 allocations: 18.06 KiB)

# sort(randn(1000)) allocates less
julia> @btime sort(x) setup=(x=randn(1000));
  23.551 μs (7 allocations: 26.00 KiB)  =>  22.088 μs (4 allocations: 14.12 KiB)

# non-IEEEFloat unions with missing are faster [fix regression]
julia> @btime sort(x) setup=(x=shuffle!(vcat(fill(missing, 10), rand(Int, 100))));
  4.265 μs (2 allocations: 2.12 KiB)  =>  2.596 μs (2 allocations: 1.94 KiB)

# IEEEFloat unions with missing are faster [fix regression]
julia> @btime sort(x) setup=(x=shuffle!(vcat(fill(missing, 10), rand(100))));
  4.322 μs (22 allocations: 4.62 KiB)  =>  2.728 μs (2 allocations: 1.94 KiB)

# non-IEEEFloat unions with missing are faster for large vectors
julia> @btime sort(x) setup=(x=shuffle!(vcat(fill(missing, 10), rand(Int, 1000))));
  60.020 μs (2 allocations: 18.12 KiB)  =>  25.735 μs (3 allocations: 19.19 KiB)

# IEEEFloat unions with missing are faster for large vectors
julia> @btime sort(x) setup=(x=shuffle!(vcat(fill(missing, 10), rand(1000))));
  56.334 μs (283 allocations: 41.72 KiB)  =>  22.828 μs (3 allocations: 19.19 KiB)

# sortperm is relatively unaffected
julia> @btime sortperm(x) setup=(x=rand(1000));
  32.822 μs (6 allocations: 23.86 KiB)  =>  32.082 μs (5 allocations: 15.92 KiB)

# sort(x; dims) is _way_ faster [fix regression]
julia> for i in 1:3 @btime sort(rand(100,100,50); dims=$i) end
  896.157 ms (10007 allocations: 18.63 GiB)  =>  13.005 ms (7 allocations: 7.63 MiB)
  953.649 ms (10022 allocations: 18.64 GiB)  =>  14.739 ms (22 allocations: 11.45 MiB)
  2.477 s (20022 allocations: 37.26 GiB)  =>  12.357 ms (21 allocations: 11.45 MiB)

# sort!(x; dims) is a bit faster
julia> for i in 1:3 @btime sort!(rand(100,100,50); dims=$i) end
  15.039 ms (15003 allocations: 8.70 MiB)  =>  13.466 ms (15003 allocations: 4.27 MiB)
  15.512 ms (15003 allocations: 8.70 MiB)  =>  14.879 ms (15003 allocations: 4.27 MiB)
  11.899 ms (20003 allocations: 5.04 MiB)  =>  10.915 ms (3 allocations: 3.82 MiB)
```

Fixes #47721
Unrelatedly, fixes #47474, closes #47383

Commit:
initial functionality

Commit:
test for bug that slipped through test suite

Commit:
fix bug

Commit:
make send_to_end more human friendly (and less compiler friendly! introduces regressions.)

Commit:
Give each sorting pass and DEFAULT_STABLE a docstring

Commit:
support 5- and 3-argument sort! for backwards compatability

petvana commented:
Nice. 

There seems to be a performance regression for `sortperm`. This may be caused by removing `fpsort!`.
``` julia
julia> @btime sortperm(x) setup=(x=rand(1000));
  29.040 μs (6 allocations: 23.86 KiB) =>   57.107 μs (5 allocations: 15.92 KiB)
```
(Btw, it seems possible to speed up `sortperm` by utilizing stable sort in future.)

Commit:
add tests and fix typos they unveiled

petvana commented on base/sort.jl:
```diff
@@ -407,111 +407,232 @@ function insorted end
+        j = send_to_end!(x -> after_zero(o, x), v; lo, hi)
+        _sort!(iv, a.next, Reverse; lo, hi=j, kw...)
+        _sort!(iv, a.next, Forward; lo=j+1, hi, kw...)
+        v
```
```suggestion
```

petvana commented on base/sort.jl:
```diff
@@ -760,133 +1071,119 @@ function _issorted(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
+Finally, if the input has length less than 80, we dispatch to [`InsertionSort`](@ref) and
+otherwise we dispatch to [`QuickSort`](@ref).
+"""
+const DEFAULT_STABLE = InitialOptimizations(IsUIntMappable(
```
This seems quite hard to read (and modify). Would it be possible to simplify it for the reader, smth like as follows?
``` julia
const DEFAULT_STABLE = begin
          alg1 = QuickSort |> Small{80} |> ConsiderRadixSort |> ConsiderCountingSort |> ComputeExtrema |> CheckSorted |> Small{40} 
          IsUIntMappable(alg1, StableCheckSorted(QuickSort)) |> InitialOptimizations
       end
```
Also the outputed algorithm is kind of unusefull for the user :-D
``` julia
julia> Base.Sort.defalg([1,2])
Base.Sort.MissingOptimization{Base.Sort.BoolOptimization{Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}(Base.Sort.BoolOptimization{Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}(Small{10, Base.Sort.InsertionSortAlg, Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}(Base.Sort.InsertionSortAlg(), Base.Sort.IEEEFloatOptimization{Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}(Base.Sort.IsUIntMappable{Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}(Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}(Base.Sort.InsertionSortAlg(), Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}(Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}(Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}(Base.Sort.CountingSort(), ConsiderRadixSort{Base.Sort.RadixSort, Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}(Base.Sort.RadixSort(), Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}(Base.Sort.InsertionSortAlg(), PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}(missing, missing, Base.Sort.InsertionSortAlg()))))))), Base.Sort.StableCheckSorted{PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}(PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}(missing, missing, Base.Sort.InsertionSortAlg())))))))
```

LilithHafner replied:
Yeah... that output is especially terrible.

LilithHafner replied:
I prefer not to reverse the order with `|>` because the standard order reflects the order of operations. `MissingOptimization(InsertionSort())` first tries to perform a missing optimization and then performs insertion sort.

I've reformatted the const declarations and added pretty printing:

```julia
const DEFAULT_STABLE = InitialOptimizations(
    IsUIntMappable(
        Small{40}(
            CheckSorted(
                ComputeExtrema(
                    ConsiderCountingSort(
                        ConsiderRadixSort(
                            Small{80}(
                                QuickSort)))))),
        StableCheckSorted(
            QuickSort)))
```
and
```
julia> Base.DEFAULT_STABLE
Base.Sort.MissingOptimization(
    Base.Sort.BoolOptimization(
        Base.Sort.Small{10}(
            InsertionSort(),
            Base.Sort.IEEEFloatOptimization(
                Base.Sort.IsUIntMappable(
                    Base.Sort.Small{40}(
                        InsertionSort(),
                        Base.Sort.CheckSorted(
                            Base.Sort.ComputeExtrema(
                                Base.Sort.ConsiderCountingSort(
                                    Base.Sort.CountingSort(),
                                    Base.Sort.ConsiderRadixSort(
                                        Base.Sort.RadixSort(),
                                        Base.Sort.Small{80}(
                                            InsertionSort(),
                                            PartialQuickSort(missing, missing,
                                                InsertionSort()))))))),
                    Base.Sort.StableCheckSorted(
                        PartialQuickSort(missing, missing,
                            InsertionSort())))))))
```
It's still fairly large, but I don't think there's any way of getting around that without either simplifying dispatch (possible a good idea for a future PR) or moving many optimization passes outside the scope of `DEFAULT_STABLE` (and therefore not easily configured by a user)

petvana replied:
> I've reformatted the const declarations and added pretty printing:

Thanks, it's definitely better. Btw, this seems like a nice case for reversed `<|` operator.

petvana commented on base/sort.jl:
```diff
@@ -603,75 +960,46 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    if _issorted(v, lo, hi, o)
+        return v
+    elseif _issorted(v, lo, hi, Lt((x, y) -> !lt(o, x, y)))
+        # Reverse only if necessary. Using issorted(..., Reverse(o)) would violate stability.
```
It seems possible to use `issorted(..., Reverse(o))` and then implement `_stablereverse!(v, lo, hi, o)` that would handle the stability such that a segment of equal elements is reversed back at the end. This all is O(n) but outside of this PR's scope.

Commit:
avoid potential name conflict

Commit:
switch to custom keyword handling

FIXES UNEXPECTED ALLOCATIONS
removes code that previously harbored bugs that slipped through the test suite

Commit:
remove InsertionSortAlg and MergeSortAlg

Commit:
better algorithm display

Commit:
stop passing U around

Fixes a few remaining unexpected allocations
U can be statically computed from the type of v and order so there is no need.
Further, U is infered as ::DataType rather than Type{U} which causes type instabilities.

Commit:
remove lenm1

it is invalid to cache lenm1 because lo and hi may be redefined
and we have no cache invalidation system

Commit:
fix unexpected allocations in Radix Sort

fixes #47474
in this PR rather than separate to avoid dealing with the merge

Commit:
fix doctests? I have no idea how

KristofferC commented:
Probably good with a PkgEval here before merging.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1306866487) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f06de10_vs_26cb6d5/report.html).

LilithHafner commented:
Breaks SortingAlgorithms which has 2033 direct + indirect deps.

Commit:
support and test backwards compatability with packages that depend in sorting internals

Commit:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/63 and [38f4512](https://github.com/JuliaLang/julia/pull/47383/commits/38f4512d18bdb96d14b4b3e14d299b5d53affcd4) fix SortingAlgorithm.jl's build and all tests except for stability of NaN when sorting with unstable algorithms. Those break because this PR makes `sort(v, SortingAlgorithms.CombSort)` perform comb sort and only comb sort, skipping IEEEFloat optimizations (that happen to stabilize NaNs). If SortingAlgorithms wants to keep those optimizations they can define `const CombSort = Base.Sort.InitialOptimizations(CombSortAlg())`.

Is it possible to tell nanosoldier to use [this branch](https://github.com/JuliaCollections/SortingAlgorithms.jl/tree/fallback-to-base-radixsort) of SortingAlgorithms or do we have to merge and release https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/63 first?

Commit:
improve extensibility tests

Commit:
overhall scratch space handling

make _sort! return scratch space rather than sorted vector
so that things like IEEEFloatOptimization can re-use the
scratch space allocated on their first recursive call

maleadt commented:
> Is it possible to tell nanosoldier to use [this branch](https://github.com/JuliaCollections/SortingAlgorithms.jl/tree/fallback-to-base-radixsort) of SortingAlgorithms or do we have to merge and release [JuliaCollections/SortingAlgorithms.jl#63](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/63) first?

You can set the `registry` configuration argument which specifies which branch on `JuliaRegistries/general` to use, see e.g. https://github.com/JuliaLang/julia/pull/45540#issuecomment-1210612971. Probably easier to tag and release SortingAlgorithms :-)

LilithHafner commented:
@nanosoldier `runtests([SortingAlgorithms], vs = ":master")`

LilithHafner commented:
@nanosoldier `runtests(["SortingAlgorithms"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1313130425) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c633419_vs_7116a8b/report.html).

LilithHafner committed:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
SortingAlgorithms.jl's tests still failed, but its build passed and the test failure is due to overly strict testing; not a problem that should affect downstream packages.

@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1313147402) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/32a6f54_vs_7116a8b/report.html).

Commit:
Consistency with other constructors

Commit:
pass around even fewer easily computed things in kw to reduce load on the compiler

LilithHafner committed:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
@nanosoldier `runtests(["DataFrames", "StatsBase"], vs = ":master")`

Commit:
revert "remove InsertionSortAlg and MergeSortAlg" for backwards compatability with folks who use internals (DataFrames.jl)

LilithHafner commented:
@nanosoldier `runtests(["DataFrames", "StatsBase"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1320844561) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/69677ba_vs_1d8f7e0/report.html).

Commit:
remove type constraint that is trickkky for the compiler to handle

Commit:
improve legacy dispatch system

LilithHafner commented:
@nanosoldier `runtests(["DataFrames", "StatsBase"], vs = ":master")`

LilithHafner committed:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
Actually, I'm optimistic based on local tests. Sorry for the stray tiny job.
@nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1321148278) has completed - no new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ab549f5_vs_90934ff/report.html).

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1321149881) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/61e8fc9_vs_90934ff/report.html).

Commit:
fix 5-arg MergeSort and add tests for 5-arg sort

Commit:
cleanup interpolation to make JET.jl happy

Commit:
fix and test handling -0.0 in IEEEFloatOptimization

LilithHafner commented:
Down to 83 failures and 13 new passes. Some of the failures were bugs in the PR which I have now fixed. Rerunning the failures:

@nanosoldier `runtests(["BayesHistogram", "ProgenyTestingTools", "Immerse", "DeformableBodies", "Firefly", "DeepDish", "GenomicVectors", "UAParser", "FilteredGroupbyMacro", "DTALib", "JSONLines", "RenderJay", "LuxLib", "LITS", "GEEBRA", "OptimKit", "RvSpectMLPlots", "PopGen", "CopEnt", "ANOVA", "CrossfilterCharts", "MultiModalMuSig", "Hadleyverse", "StanDataFrames", "ClimateTasks", "GroupedErrors", "MLJJLBoost", "ModelConstructors", "PkgDependency", "ParticleMDI", "SoilProfiles", "DynamicHMC", "SortingAlgorithms", "ReferenceTests", "JET", "KeywordSearch", "ChemometricsTools", "EvaluationCF", "DevIL", "EchogramImages", "ManifoldsBase", "Phylo", "ChemometricsData", "JuliaDBMeta", "DPClustering", "PhyloNetworks", "OpSel", "Schrute", "ARules", "LockandKeyLookups", "RigorousCoupledWaveAnalysis", "MotifRecognition", "Hydrographs", "QuerySQLite", "ArgoData", "CSVReader", "MaximumLikelihoodProblems", "MaxEntropyGraphs", "JLBoostMLJ", "OMOPCommonDataModel", "InvariantCausal", "OptimalTransmissionRouting", "BlockDiagonals", "REDCap", "CORBITS", "OhMyQSIM", "PosDefManifoldML", "StructArrays", "KeyedFrames", "InMemoryDatasets", "LifeTable", "TopologyPreprocessing", "SantasLittleHelpers", "Mex", "AppliGeneralLedger", "TheCannon", "MultiDimEquations", "GlobalSearchRegression", "JLBoost", "MatrixEquations", "StanMamba", "MLJTuning"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1321456274) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/fef85c0_vs_c9eccfc/report.html).

Commit:
fix and test bug where countsort's correct overflow behavior triggers error due to unexpected promotion to UInt

Commit:
add type signature to reduce possible method ambiguities (e.g. with AbstractTrees v0.3.4's ImplicitRootState)

Commit:
support 6-argument sort! because people do actually use it.

LilithHafner committed:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
Down to three real or semi-real errors (now fixed) and a whole bunch of packages that depend on SortingAlgorithms 0.3 (and therefore don't get the fix in 1.1.0). SortingAlgorithms 1.0 was released over a year ago and is, ironically, a [non breaking release](https://github.com/JuliaCollections/SortingAlgorithms.jl/commit/aa2b98d384ddd132aae0219e68fb63b92513cb35#commitcomment-50066397). Literally nobody should be compat bounded to 0.3. However, some packages are on an old version of DataFrames that predates SortingAlgorithms 1.0 and therefore cannot use SortingAlgorithms 1.0 unless we retroactively expand the compat of released package versions or they deal with breaking changes in newer versions of DataFrames.

@nanosoldier `runtests(ALL, vs=:master)`

maleadt commented:
That @nanosoldier invocation syntax is wrong, the `vs` should be quoted. Even better; drop it at all, as it isn't required anymore:

@nanosoldier `runtests(ALL)`

LilithHafner commented:
I'd like to merge this and backport to 1.9.

If we don't backport this to 1.9, then most sorting backports will need to be manual and all should be manual to avoid semantic conflicts. I would not enjoy simultaneously supporting two very different sorting internals at the same time.

CI has been green for a while, and PkgEval is okay but not great. Many (though likely not all) of the issues are downstream and I am working on fixing them (https://github.com/JuliaCollections/SortingAlgorithms.jl/commit/f9d60839541fecd733041705e902fd309016b5fb), but I'd like to merge this now/before the next 1.9-alpha release so that it can get more feedback before the full 1.9 release and so that we can start working on PRs that are waiting for this (e.g. #47715, #47587)

@KristofferC, I know you said "Probably good with a PkgEval here before merging." and PkgEval is still not totally clean. Is this plan acceptable to you and anyone else who is interested?

LilithHafner commented:
Bump, this is an awfully large PR and it's likely to accumulate merge conflicts fairly quickly.

LilithHafner committed:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
Rerunning failed PkgEval now that SortingAlgorithms 3.2 has been released to work around compat bounds restricting to below the non-breaking 1.0 release.

@nanosoldier `runtests(["ARules", "JuliaDBMeta", "GEEBRA", "ParticleMDI", "MultiModalMuSig", "SortingAlgorithms", "SoilProfiles", "DeepDish", "LazyAlgebra", "GenomicVectors", "JLBoost", "TopologyPreprocessing", "MaxEntropyGraphs", "ChemometricsTools", "StaticTools", "RigorousCoupledWaveAnalysis", "DeformableBodies", "LockandKeyLookups", "RvSpectMLPlots", "HydrophoneCalibrations", "Minio", "CSVReader", "ModelConstructors", "SantasLittleHelpers", "EvaluationCF", "MLJJLBoost", "StanMamba", "ArgoData", "NeuralQuantumState", "Mex", "JLBoostMLJ", "DTALib", "ANOVA", "NumericalAlgorithms", "InvariantCausal", "RenderJay", "FilteredGroupbyMacro", "MaximumLikelihoodProblems", "REDCap", "OpSel", "UAParser", "PkgDependency", "RigidBodyTools", "TheCannon", "ProgenyTestingTools", "ChemometricsData", "AppliGeneralLedger", "PosDefManifoldML", "VariantCallFormat", "Schrute", "Whitebox", "CORBITS", "KeyedFrames", "ReferenceTests", "Recommenders", "Hadleyverse", "QuerySQLite", "MotifRecognition", "OptimalTransmissionRouting", "OMOPCommonDataModel", "LITS", "EchogramImages", "Phylo", "CrossfilterCharts", "DPClustering", "ClimateTasks", "StanDataFrames", "SparseRegression", "JSONLines", "LifeTable", "Hydrographs", "Firefly", "Immerse", "GlobalSearchRegression", "Infernal"])`

LilithHafner commented:
Rerunning failed PkgEval now that SortingAlgorithms 3.2 has been released to work around compat bounds restricting to below the non-breaking 1.0 release.

@nanosoldier `runtests(["ARules", "JuliaDBMeta", "GEEBRA", "ParticleMDI", "MultiModalMuSig", "SortingAlgorithms", "SoilProfiles", "DeepDish", "LazyAlgebra", "GenomicVectors", "JLBoost", "TopologyPreprocessing", "MaxEntropyGraphs", "ChemometricsTools", "StaticTools", "RigorousCoupledWaveAnalysis", "DeformableBodies", "LockandKeyLookups", "RvSpectMLPlots", "HydrophoneCalibrations", "Minio", "CSVReader", "ModelConstructors", "SantasLittleHelpers", "EvaluationCF", "MLJJLBoost", "StanMamba", "ArgoData", "NeuralQuantumState", "Mex", "JLBoostMLJ", "DTALib", "ANOVA", "NumericalAlgorithms", "InvariantCausal", "RenderJay", "FilteredGroupbyMacro", "MaximumLikelihoodProblems", "REDCap", "OpSel", "UAParser", "PkgDependency", "RigidBodyTools", "TheCannon", "ProgenyTestingTools", "ChemometricsData", "AppliGeneralLedger", "PosDefManifoldML", "VariantCallFormat", "Schrute", "Whitebox", "CORBITS", "KeyedFrames", "ReferenceTests", "Recommenders", "Hadleyverse", "QuerySQLite", "MotifRecognition", "OptimalTransmissionRouting", "OMOPCommonDataModel", "LITS", "EchogramImages", "Phylo", "CrossfilterCharts", "DPClustering", "ClimateTasks", "StanDataFrames", "SparseRegression", "JSONLines", "LifeTable", "Hydrographs", "Firefly", "Immerse", "GlobalSearchRegression", "Infernal"])`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1333671261) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/report.html).

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1333673753) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/report.html).

LilithHafner commented:
A total of 5 PkgEval failures remain, three of which have identified non-blocking causes and two of which seem unrelated but have no known cause.

Package has test failures (3 packages):
Minio v0.2.0: [fail](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/Minio.primary.log) vs. [ok](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/Minio.against.log) - Networking

Recommenders v0.1.2: [fail](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/Recommenders.primary.log) vs. [ok](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/Recommenders.against.log) - Error while sorting a SparseArray that previously produced a dense result. There is an open PR to fix this, but a maintainer has indicated that they prefer an error in this case, so I do not view this as a regression.

SortingAlgorithms v1.1.0: [fail](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/SortingAlgorithms.primary.log) vs. [ok](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/SortingAlgorithms.against.log) - Due to testing that HeapSort, and unstable sorting algorithm, preserves the order of input NaNs. We can loosen the test or adjust the hook into Julia's internals to maintain this property, but either way it is not a behavior that anyone downstream depends on (at least not in PkgEval).

There were unidentified errors (1 packages):
JSONLines v2.0.1: [fail](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/JSONLines.primary.log) vs. [ok](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1ea32fc_vs_04a4edb/JSONLines.against.log) - Seems entirely unrelated; rerunning

I'm going to investigate the final failure further, but **unless I find a regression related to this PR or anyone objects, I plan to merge this soon.**

[This is the diff](https://github.com/JuliaLang/julia/pull/47383/files/f06de1089edc3fc261875f201782f171dcecdcd9..1ea32fc0d791f82dde3234fc94ede887ea19645c) since @KristoferC's ["Probably good with a PkgEval here before merging."](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1306748946)

It contains added tests and bug fixes from PkgEval, simplifications, [renaming buffer => scratch](47172), and better scratch space handling to architecture to reduce allocations and support future improvements.

Rerunning the package whose test failure confuses me: `@nanosoldier runtests(["JSONLines"])`

LilithHafner committed:
Merge branch 'master' into sort-dispatch

LilithHafner commented:
Rerunning the package whose test failure confuses me: @nanosoldier `runtests(["JSONLines"])`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1334584503) has completed - no new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/3d0a502_vs_cc25a13/report.html).

LilithHafner commented:
I also couldn't reproduce that package's test failures locally, so the LGTM

LilithHafner commented:
I expect this to fail; though it worked on the prior commit: @nanosoldier runbenchmarks("sort")

LilithHafner commented:
I expect this to fail; though it worked on the prior commit: @nanosoldier `runbenchmarks("sort")`

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/47383#issuecomment-1340864563) failed.

LilithHafner commented:
This PR broke `runbenchmarks("sort")`; https://github.com/JuliaLang/julia/pull/47822 fixes it.

vtjnash commented:
Did this end up getting any nanosoldier testing? `quicksort` performance seemed to do very poorly on the latest nightly, which seems likely this commit to be the culprit

LilithHafner commented:
Just looked into this and the reason we have regressions is that this PR makes `QuickSort` specify using exactly and only the quicksort sorting method. Notably, this excludes floating pojnt optimizations and presorted checks which results in huge regressions for presorted floating point inputs (including ascending, descending, and all ones, which is also sorted) and substantial regressions for random floating point inputs.

We didn't catch these regression earlier because they only effect usage that explicitly specifies `alg=QuickSort`, `alg=MergeSort`, etc. and the more worrying regression only comes up for floating points. BaseBenchmarks benchmarks (only) the use case of sorting floating point vectors with explicitly specified algorithms. After https://github.com/JuliaCI/BaseBenchmarks.jl/pull/305, it will continue to benchmark that case and also other cases.
